Assembler.UnitTests/InitialMessageHandlerTests.cs

```diff
@@ -0,0 +1,182 @@
+
+            _handler = new InitialMessageHandler<BaseFrame, BaseMessageInAssembly>(_cacheMock.Object,
+                _identifierFactoryMock.Object, _messageInAssemblyCreatorMock.Object, _enricherMock.Object,
+                Utilities.GetLoggerFactory());
+
+            _handler.MessageAssemblyFinished += _assembledMessages.Add;
+        }
+
+        [TearDown]
+        public void Teardown()
+        {
+            _identifierFactoryMock.VerifyNoOtherCalls();
+            _enricherMock.VerifyNoOtherCalls();
+            _cacheMock.VerifyNoOtherCalls();
+            _messageInAssemblyCreatorMock.VerifyNoOtherCalls();
+        }
+
+        [Test]
+        public void Handle_IdentifierThrowsException_FrameNotBeingUsed()
+        {
+            // Arrange
+            var frame = new Mock<BaseFrame>(FrameType.Initial);
+
+            _identifierFactoryMock.Setup(identifier => identifier.Create(It.IsAny<BaseFrame>()))
+                .Throws<NullReferenceException>();
+
+            // Act
+            _handler.Handle(frame.Object);
+
+            // Assert
+            _identifierFactoryMock.Verify(identifier => identifier.Create(It.IsAny<BaseFrame>()), Times.Once);
+            _identifierFactoryMock.Verify(identifier => identifier.Create(frame.Object), Times.Once);
+        }
+
+        [Test]
+        public void Handle_MessageInCacheAndMiddleReceived_ReleasesTheOldOneAndCreatesANewOne()
+        {
+            // Arrange
+            var frame = new Mock<BaseFrame>(FrameType.Initial);
+            var message = new Mock<BaseMessageInAssembly>();
+            var newMessage = new Mock<BaseMessageInAssembly>();
+            message.Object.MiddleReceived = true;
+
+            _cacheMock.Setup(cache => cache.Exists(It.IsAny<string>())).Returns(true);
+            _cacheMock.Setup(cache => cache.Get<BaseMessageInAssembly>(It.IsAny<string>())).Returns(message.Object);
+            _messageInAssemblyCreatorMock.Setup(creator => creator.Create()).Returns(newMessage.Object);
+
+            // Act
+            _handler.Handle(frame.Object);
+
+            // Assert
+            _identifierFactoryMock.Verify(identifier => identifier.Create(It.IsAny<BaseFrame>()), Times.Once);
+            _identifierFactoryMock.Verify(identifier => identifier.Create(frame.Object), Times.Once);
+
+            _cacheMock.Verify(cache => cache.Exists(It.IsAny<string>()), Times.Once);
+            _cacheMock.Verify(cache => cache.Exists(_identifierString), Times.Once);
+
+            _cacheMock.Verify(cache => cache.Get<It.IsAnyType>(It.IsAny<string>()), Times.Once);
+            _cacheMock.Verify(cache => cache.Get<It.IsAnyType>(_identifierString), Times.Once);
+
+            _cacheMock.Verify(cache => cache.Remove(It.IsAny<string>()), Times.Once);
+            _cacheMock.Verify(cache => cache.Remove(_identifierString), Times.Once);
+
+            Assert.AreEqual(ReleaseReason.AnotherMessageStarted, message.Object.ReleaseReason);
```

Comment:
I think you should move the assertion to the end of the test after all the verify

Reply:
But that's the order that things happen